### PR TITLE
DR2-1863 Upsert to handle multiple trees.

### PR DESCRIPTION
**Pull request by MancunianSam**
The way this works now is:

1 Get all the archive folder rows from Dynamo, sorted by parent path, shortest first
2 Get a map if sourceId name -> entity
3 For every row in order, foldLeft with a map of DynamoId -> EntityId
  a) Check if it is in the map from 2 and return entity ref if it is
  b) If it is, check the map from 3 for a parent and add it to Preservica
4. Find the rows which already existed by filtering the rows from 1 with
   the keys from 2
5. Do the update identifiers and entities as before.

### CODE REVIEW

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala [hunk 1]
```diff
@@ -6,83 +6,149 @@ import io.circe.generic.auto.*
...
+    def getFolderRows: IO[List[ArchiveFolderDynamoTable]] =
+      dependencies.dADynamoDBClient
+        .getItems[ArchiveFolderDynamoTable, FilesTablePrimaryKey](
+          folderIdPartitionKeysAndValues,
+          config.archiveFolderTableName
+        )
+        .map(_.sorted)
+
+    def getIdToEntityMap(folderRows: List[ArchiveFolderDynamoTable], sourceMap: Map[String, Entity]) =
+      folderRows.foldLeft(IO(Map.empty[UUID, UUID])) { (entityMapIO, folderRow) =>
```
Comment (techncl) [inline scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:51]: Maybe `idToEntityRefMapIO`?

**Review by techncl — COMMENTED** [review]
Looks good just some minor comments. Also, it seems as though some of the checks that I added previously are missing.

1. How come you've removed the check that makes sure that parent paths increment by 1? I know there are multiple trees now, but couldn't we still do that (by maybe grouping folders by BatchId or something?
2. Same question for the `checkEachParentPathMatchesFolderBeforeIt` check

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala [hunk 2]
```diff
@@ -6,83 +6,149 @@ import io.circe.generic.auto.*
...
-        dependencies.dADynamoDBClient,
-        folderIdPartitionKeysAndValues,
-        config.archiveFolderTableName
-      )
-
-      _ <- checkNumOfParentPathSlashesPerFolderIncrease(folderRowsSortedByParentPath)
-      _ <- checkEachParentPathMatchesFolderBeforeIt(folderRowsSortedByParentPath)
-
-      secretName = config.secretName
+    def getFolderRows: IO[List[ArchiveFolderDynamoTable]] =
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:42]: It's used quite a bit and that name is quite long, it makes the lines quite long. I'll add the implicit ordering to the `getFolderRows` method definition so it's clear that the method is sorting.

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala [hunk 3]
```diff
@@ -6,83 +6,149 @@ import io.circe.generic.auto.*
...
+          .map { entity =>
+            val expectedParent = getExpectedParent(folderRow, folderRows, sourceMap)
+            for {
+              _ <- IO.raiseWhen(!entity.entityType.contains(StructuralObject))(
+                new Exception(s"The entity type for folder id ${folderRow.id} should be 'StructuralObject' but it is ${entity.entityType.toStringOrEmpty}")
+              )
+              _ <- IO.raiseWhen(entity.parent != expectedParent)(
+                new Exception(s"API returned a parent ref of '${entity.parent.toStringOrEmpty}' for entity ${entity.ref} instead of expected '${expectedParent.toStringOrEmpty}'")
+              )
+              _ <- IO.raiseWhen(entity.securityTag.isEmpty || entity.securityTag.flatMap(Option.apply).isEmpty)(
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:63]: The Preservica client will return `None` if the security tag is anything other than `Open` or `Closed` so this covers that.

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala [hunk 4]
```diff
@@ -6,83 +6,149 @@ import io.circe.generic.auto.*
...
+        }
+    }
 
-      _ <- folderInfoOfEntitiesThatExistWithSecurityTags.map { fi =>
-        val identifiersFromDynamo = fi.folderRow.identifiers.map(id => Identifier(id.identifierName, id.value))
+    val logWithBatchRef = log(Map("batchRef" -> stepFnInput.batchId))(_)
 
-        val entity = fi.entity.get
+    for {
+      folderRows <- getFolderRows
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:132]: It's a bit long again and it is used quite a bit so it makes it quite hard to parse. I've added an implicit ordering to the method definition so you can see it's sorting something.

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala [hunk 5]
```diff
@@ -6,83 +6,149 @@ import io.circe.generic.auto.*
...
+      entitiesByIdentifier <- getEntitiesByIdentifier(folderRows)
+      idToEntityRef <- getIdToEntityMap(folderRows, entitiesByIdentifier)
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:134]: Good point. I've made the method return `IO[Unit]` and changed it to `createNonExistentFolders`

**Comment by MancunianSam** [comment]
I removed the check counting the paths because I couldn't get it to work with two completely seperate file trees which we can have. We can have files with paths that are completely unrelated so there's no way I can see to group them. Same with `checkEachParentPathMatchesFolderBeforeIt` 

I haven't removed any tests though and they're still passing with slightly different error messages so we are still validating the same things, just not in the same way.

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala [hunk 6]
```diff
@@ -39,46 +39,47 @@ class Lambda extends LambdaRunner[StepFnInput, Unit, Config, Dependencies] {
...
+              } yield updatedEntityMap
+            }
+            .getOrElse {
+              for {
+                entityMap <- idToEntityRefMapIO
+                entityId <- addFolder(folderRow, entityMap)
+              } yield entityMap + (folderRow.id -> entityId)
+            }
+        }
+        .map(_ => ())
```
Comment (techncl) [inline scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:77]: You can just do `.void` here

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala [hunk 7]
```diff
@@ -6,83 +6,149 @@ import io.circe.generic.auto.*
...
-      _ <- checkEachParentPathMatchesFolderBeforeIt(folderRowsSortedByParentPath)
-
-      secretName = config.secretName
+    def getFolderRows: IO[List[ArchiveFolderDynamoTable]] =
+      dependencies.dADynamoDBClient
+        .getItems[ArchiveFolderDynamoTable, FilesTablePrimaryKey](
+          folderIdPartitionKeysAndValues,
+          config.archiveFolderTableName
+        )
+        .map(_.sorted)
```
[the remark this answers is not in the document]
Comment (techncl) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:48]: Ahh, I see. Should we add a test for this (to ensure that they are added in this order)? Because someone in the future could remove `.map(_.sorted)`, thinking it's not necessary.

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (techncl) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:133]: Ok. Maybe we could also give the `given` a name?

**Review by techncl — COMMENTED** [review]
👍, just a few more comments. Also, I forgot to add previously, could we add some basic steps to the README?

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:48]: I tried removing the `.map(_.sorted)` and one of the tests failed so I think we're already covered.

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala (hunk 6 above)
Comment (MancunianSam) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:77]: Nice, thanks.

File: scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (MancunianSam) [reply scala/lambdas/ingest-upsert-archive-folders/src/main/scala/uk/gov/nationalarchives/ingestupsertarchivefolders/Lambda.scala:133]: It's not really necessary so I'll leave it. As I say, I'd like to move a bit more towards using the types to see what a method is doing rather than naming everything explicitly

**Comment by MancunianSam** [comment]
I've added in a section to the README to show how it works.
